### PR TITLE
Fix MachBuffer branch handling with redirect chains.

### LINKED ISSUE #2082: Cranelift: broken two-way jump in Godot engine demo
This has taken some time to track down, but I eventually managed to pinpoint to one function that was sufficient to trigger the crash in Firefox, and so I went on and ran Firefox in a debugger, stepping in until i could see where control flow seemed to go wrong.

Here's a summary of the input function, plus a runtime trace to explicit where the bug is (and where i observed the crash):

<details><summary>Clif test</summary>

```
function u0:7705(i32, i64 vmctx) baldrdash_system_v {
    gv0 = vmctx
    gv1 = load.i64 notrap aligned readonly gv0
    gv2 = iadd_imm.i64 gv0, 56
    heap0 = static gv1, min 0x0100_0000, bound 0x0001_0000_0000, offset_guard 0x8000_0000, index_type i32
    jt0 = jump_table [block14, block13]

block0(v0: i32, v1: i64):
    v31 -> v0
    v32 -> v31
    v2 = iconst.i32 0
    v3 = heap_addr.i64 heap0, v0, 1
    v4 = sload16.i32 v3+420
    v5 = iconst.i32 0xffff
    v6 = band v4, v5
    v14 -> v6
    v29 -> v14
    v7 = iconst.i32 524
    v8 = iadd v0, v7
    v17 -> v8
    v30 -> v17
    brz v4, block15
    jump block6

block6:
    v9 = iconst.i32 0
    v10 = iconst.i32 0
    jump block7(v9, v10)

block7(v13: i32, v52: i32):
    v26 -> v13
    v11 = global_value.i64 gv2
    v12 = load.i32 notrap aligned v11
    resumable_trapnz v12, interrupt
    v15 = icmp sge v13, v14
    v16 = bint.i32 v15
    brnz v16, block4(v31, v52, v14, v17)
    jump block10

block10:
    v18 = heap_addr.i64 heap0, v17, 1
    v19 = load.i32 v18
    v20 = iadd v19, v13
    v21 = heap_addr.i64 heap0, v20, 1
    v22 = sload8.i32 v21
    v23 = iconst.i32 -1
    v24 = isub v22, v23
    br_table v24, block12, jt0

block14:
    jump block11(v52)

block13:
    v25 = iconst.i32 1
    jump block11(v25)

block12:
    jump block9

block11(v53: i32):
    v27 = iconst.i32 1
    v28 = iadd.i32 v13, v27
    jump block7(v28, v53)

block9:
    jump block8

block8:
    v33 = heap_addr.i64 heap0, v0, 1
    v34 = load.i32 v33+216
    jump block5

block15:
    v35 = iconst.i32 0
    jump block4(v0, v35, v6, v8)

block5:
    jump block3(v34, v0, v14, v17)

block4(v36: i32, v51: i32, v111: i32, v114: i32):
    v65 -> v36
    v110 -> v111
    v113 -> v114
    v37 = iconst.i32 216
    v38 = iadd v36, v37
    v39 = heap_addr.i64 heap0, v38, 1
    v40 = load.i32 v39
    v41 = iconst.i32 0xffff_ffff_ff7f_ffff
    v42 = band v40, v41
    v43 = heap_addr.i64 heap0, v38, 1
    store v42, v43
    v44 = iconst.i32 212
    v45 = iadd v36, v44
    v46 = heap_addr.i64 heap0, v45, 1
    v47 = load.i32 v46
    v48 = iconst.i32 -8193
    v49 = band v47, v48
    v50 = heap_addr.i64 heap0, v45, 1
    store v49, v50
    brz v51, block18
    jump block17

block17:
    jump block16(v42)

block18:
    v54 = iconst.i32 0xffff_ffff_ff7f_fe7f
    v55 = band.i32 v40, v54
    v56 = heap_addr.i64 heap0, v38, 1
    store v55, v56
    jump block16(v55)

block16(v58: i32):
    jump block3(v58, v65, v110, v113)

block3(v57: i32, v64: i32, v109: i32, v112: i32):
    v95 -> v109
    v98 -> v112
    v59 = iconst.i32 4352
    v60 = band v57, v59
    v61 = iconst.i32 4352
    v62 = icmp eq v60, v61
    v63 = bint.i32 v62
    brz v63, block20
    jump block21

block21:
    v66 = heap_addr.i64 heap0, v64, 1
    v67 = load.i32 v66+408
    v68 = heap_addr.i64 heap0, v64, 1
    v69 = uload8.i32 v68+448
    v70 = iconst.i32 3
    v71 = imul v69, v70
    v72 = iadd v67, v71
    v73 = heap_addr.i64 heap0, v72, 1
    v74 = uload8.i32 v73
    v75 = heap_addr.i64 heap0, v64, 1
    istore16 v74, v75+450
    v76 = iconst.i32 3
    v77 = imul v69, v76
    v78 = iadd v67, v77
    v79 = heap_addr.i64 heap0, v78, 1
    v80 = uload8.i32 v79+1
    v81 = heap_addr.i64 heap0, v64, 1
    istore16 v80, v81+452
    v82 = iadd v67, v77
    v83 = heap_addr.i64 heap0, v82, 1
    v84 = uload8.i32 v83+2
    v85 = heap_addr.i64 heap0, v64, 1
    istore16 v84, v85+454
    v86 = iconst.i32 0x0208_0000
    v87 = band.i32 v57, v86
    v88 = iconst.i32 0x0008_0000
    v89 = icmp eq v87, v88
    v90 = bint.i32 v89
    brz v90, block22
    jump block23

block23:
    v91 = iconst.i32 0
    jump block24(v91)

block24(v94: i32):
    v92 = global_value.i64 gv2
    v93 = load.i32 notrap aligned v92
    resumable_trapnz v93, interrupt
    v96 = icmp eq v94, v95
    v97 = bint.i32 v96
    brnz v97, block19
    jump block26

block26:
    v99 = heap_addr.i64 heap0, v98, 1
    v100 = load.i32 v99
    v101 = iadd v100, v94
    v102 = heap_addr.i64 heap0, v101, 1
    v103 = sload8.i32 v102
    v104 = iconst.i32 -1
    v105 = bxor v103, v104
    v106 = heap_addr.i64 heap0, v101, 1
    istore8 v105, v106
    v107 = iconst.i32 1
    v108 = iadd.i32 v94, v107
    jump block24(v108)

block22:
    jump block20

block20:
    jump block19

block19:
    jump block2

block2:
    jump block1

block1:
    fallthrough_return
}
```
</details>

<details><summary>x64 trace</summary>

don't pay attention to the codegen inefficiencies (yet)

```
VCode_ShowWithRRU {{
  Entry block: 0
Block 0:
  (original IR block: block0)
  (successor: Block 1)
  (successor: Block 3)
  (instruction range: 0 .. 14)
  Inst 0:   pushq   %rbx
  Inst 1:   pushq   %r12
  Inst 2:   pushq   %r13
  Inst 3:   virtual_sp_offset_adjust 24
  Inst 4:   movl    %edi, %ebx // edi = ebx = 0xbb0508 @ 0xbb0668
  Inst 5:   movq    %r15/pinned, %r12 // r15 = r12 = 0x7ffe52af0000
  Inst 6:   addq    %rbx, %r12 // r12 = 0x7ffe536a0508
  Inst 7:   movswq  420(%r12), %rsi // rsi = 0
  Inst 8:   movq    %rsi, %r12 // r12 = 0
  Inst 9:   andl    $65535, %r12d // 0
  Inst 10:   movq    %rdi, %r13 // rdi = r13 = 0xbb0508
  Inst 11:   addl    $524, %r13d // 0xbb0714 @ bb0874
  Inst 12:   cmpl    $0, %esi // 0
  Inst 13:   jz      taken=MachLabel(1) not_taken=MachLabel(3) --> jmp 1
Block 1:
  (original IR block: block15)
  (successor: Block 2)
  (instruction range: 14 .. 15)
  Inst 14:   jmp     MachLabel(2) --> jmp 2
Block 2:
  (successor: Block 11)
  (instruction range: 15 .. 17)
  Inst 15:   movl    $0, %edx // edx = 0
  Inst 16:   jmp     MachLabel(11) --> jmp 11
Block 3:
  (original IR block: block6)
  (successor: Block 4)
  (instruction range: 17 .. 21)
  Inst 17:   movq    %r14, %rsi
  Inst 18:   addq    $56, %rsi
  Inst 19:   movl    %r13d, %eax
  Inst 20:   jmp     MachLabel(4)
Block 4:
  (successor: Block 5)
  (instruction range: 21 .. 24)
  Inst 21:   movl    $0, %ecx
  Inst 22:   movl    $0, %edx
  Inst 23:   jmp     MachLabel(5)
Block 5:
  (original IR block: block7)
  (successor: Block 6)
  (successor: Block 7)
  (instruction range: 24 .. 27)
  Inst 24:   movl    0(%rsi), %r8d
  Inst 25:   cmpl    $0, %r8d
  Inst 26:   jz      taken=MachLabel(6) not_taken=MachLabel(7)
Block 6:
  (successor: Block 9)
  (instruction range: 27 .. 28)
  Inst 27:   jmp     MachLabel(9)
Block 7:
  (original IR block: block27)
  (successor: Block 8)
  (instruction range: 28 .. 30)
  Inst 28:   ud2 interrupt
  Inst 29:   jmp     MachLabel(8)
Block 8:
  (successor: Block 9)
  (instruction range: 30 .. 31)
  Inst 30:   jmp     MachLabel(9)
Block 9:
  (original IR block: block28)
  (successor: Block 10)
  (successor: Block 17)
  (instruction range: 31 .. 35)
  Inst 31:   cmpl    %r12d, %ecx
  Inst 32:   setnl   %r8b
  Inst 33:   cmpb    $0, %r8b
  Inst 34:   jnz     taken=MachLabel(10) not_taken=MachLabel(17)
Block 10:
  (successor: Block 11)
  (instruction range: 35 .. 36)
  Inst 35:   jmp     MachLabel(11)
Block 11:
  (original IR block: block4)
  (successor: Block 12)
  (successor: Block 14)
  (instruction range: 36 .. 59)
  Inst 36:   movq    %rdi, %rbx // 0xbb0508 @ bb0668
  Inst 37:   addl    $216, %ebx // 0xbb05e0
  Inst 38:   movl    %ebx, %eax // 0xbb05e0
  Inst 39:   movq    %r15/pinned, %rbx // 7ffe52af0000 = HEAP_BASE
  Inst 40:   addq    %rax, %rbx // 7ffe536a05e0
  Inst 41:   movl    0(%rbx), %esi // 0x2001004
  Inst 42:   movq    %rsi, %rbx // 0x2001004
  Inst 43:   andl    $-8388609, %ebx // same
  Inst 44:   movq    %r15/pinned, %rcx // HEAP_BASE
  Inst 45:   addq    %rax, %rcx // 7ffe536a05e0
  Inst 46:   movl    %ebx, 0(%rcx) // 0x2001004 -> *$rcx
  Inst 47:   movq    %rdi, %rcx // bb0508
  Inst 48:   addl    $212, %ecx // bb05dc // bb073c
  Inst 49:   movl    %ecx, %ecx // same
  Inst 50:   movq    %r15/pinned, %r8
  Inst 51:   addq    %rcx, %r8 // 7ffe536a05dc
  Inst 52:   movl    0(%r8), %r8d // read: 0x304000
  Inst 53:   andl    $-8193, %r8d // same
  Inst 54:   movq    %r15/pinned, %r9 // HEAP_BASE
  Inst 55:   addq    %rcx, %r9 // bb05dc
  Inst 56:   movl    %r8d, 0(%r9)
  Inst 57:   cmpl    $0, %edx // 0
  Inst 58:   jz      taken=MachLabel(12) not_taken=MachLabel(14) --> jmp 12
Block 12:
  (original IR block: block18)
  (successor: Block 13)
  (instruction range: 59 .. 64)
  Inst 59:   andl    $-8388993, %esi // 2001004
  Inst 60:   movq    %r15/pinned, %rbx // HEAP_BASE
  Inst 61:   addq    %rax, %rbx // 7ffe536a05e0
  Inst 62:   movl    %esi, 0(%rbx) // esi = 2001004, stored
  Inst 63:   jmp     MachLabel(13) -> jmp 13
Block 13:
  (successor: Block 16)
  (instruction range: 64 .. 65)
  Inst 64:   jmp     MachLabel(16) -> jmp 16
Block 14:
  (original IR block: block17)
  (successor: Block 15)
  (instruction range: 65 .. 66)
  Inst 65:   jmp     MachLabel(15)
Block 15:
  (successor: Block 16)
  (instruction range: 66 .. 68)
  Inst 66:   movq    %rbx, %rsi
  Inst 67:   jmp     MachLabel(16)
Block 16:
  (original IR block: block16)
  (successor: Block 23)
  (instruction range: 68 .. 70)
  Inst 68:   movq    %rsi, %rbx 2001004
  Inst 69:   jmp     MachLabel(23) --> jmp 23
Block 17:
  (original IR block: block10)
  (successor: Block 18)
  (successor: Block 41)
  (successor: Block 43)
  (instruction range: 70 .. 81)
  Inst 70:   movq    %r15/pinned, %r8
  Inst 71:   addq    %rax, %r8
  Inst 72:   movl    0(%r8), %r8d
  Inst 73:   addl    %ecx, %r8d
  Inst 74:   movl    %r8d, %r8d
  Inst 75:   movq    %r15/pinned, %r9
  Inst 76:   addq    %r8, %r9
  Inst 77:   movsbq  0(%r9), %r8
  Inst 78:   addl    $1, %r8d
  Inst 79:   cmpl    $2, %r8d
  Inst 80:   br_table %r8
Block 18:
  (original IR block: block12)
  (successor: Block 19)
  (instruction range: 81 .. 82)
  Inst 81:   jmp     MachLabel(19)
Block 19:
  (original IR block: block9)
  (successor: Block 20)
  (instruction range: 82 .. 83)
  Inst 82:   jmp     MachLabel(20)
Block 20:
  (original IR block: block8)
  (successor: Block 21)
  (instruction range: 83 .. 87)
  Inst 83:   movq    %r15/pinned, %rsi
  Inst 84:   addq    %rbx, %rsi
  Inst 85:   movl    216(%rsi), %ebx
  Inst 86:   jmp     MachLabel(21)
Block 21:
  (original IR block: block5)
  (successor: Block 22)
  (instruction range: 87 .. 88)
  Inst 87:   jmp     MachLabel(22)
Block 22:
  (successor: Block 23)
  (instruction range: 88 .. 89)
  Inst 88:   jmp     MachLabel(23)
Block 23:
  (original IR block: block3)
  (successor: Block 24)
  (successor: Block 25)
  (instruction range: 89 .. 95)
  Inst 89:   movq    %rbx, %rsi // 2001004
  Inst 90:   andl    $4352, %esi // esi = 0x1000
  Inst 91:   cmpl    $4352, %esi
  Inst 92:   setz    %sil // 0
  Inst 93:   cmpb    $0, %sil
  Inst 94:   jz      taken=MachLabel(24) not_taken=MachLabel(25) <-- BUG HERE. Instead of jumping to 24, it will jump to block 39.
Block 24:
  (successor: Block 28)
  (instruction range: 95 .. 96)
  Inst 95:   jmp     MachLabel(28)
Block 25:
  (original IR block: block21)
  (successor: Block 26)
  (successor: Block 29)
  (instruction range: 96 .. 129)
  Inst 96:   movl    %edi, %esi
  Inst 97:   movq    %r15/pinned, %rdi
  Inst 98:   addq    %rsi, %rdi
  Inst 99:   movl    408(%rdi), %edi
  Inst 100:   movq    %r15/pinned, %rax
  Inst 101:   addq    %rsi, %rax
  Inst 102:   movzbq  448(%rax), %rax
  Inst 103:   imull   $3, %eax
  Inst 104:   addl    %eax, %edi
  Inst 105:   movl    %edi, %edi
  Inst 106:   movq    %r15/pinned, %rax
  Inst 107:   addq    %rdi, %rax
  Inst 108:   movzbq  0(%rax), %rax
  Inst 109:   movq    %r15/pinned, %rcx
  Inst 110:   addq    %rsi, %rcx
  Inst 111:   movw    %ax, 450(%rcx)
  Inst 112:   movq    %r15/pinned, %rax
  Inst 113:   addq    %rdi, %rax
  Inst 114:   movzbq  1(%rax), %rax
  Inst 115:   movq    %r15/pinned, %rcx
  Inst 116:   addq    %rsi, %rcx
  Inst 117:   movw    %ax, 452(%rcx)
  Inst 118:   movq    %r15/pinned, %rax
  Inst 119:   addq    %rdi, %rax
  Inst 120:   movzbq  2(%rax), %rdi
  Inst 121:   movq    %r15/pinned, %rax
  Inst 122:   addq    %rsi, %rax
  Inst 123:   movw    %di, 454(%rax)
  Inst 124:   andl    $34078720, %ebx
  Inst 125:   cmpl    $524288, %ebx
  Inst 126:   setz    %bl
  Inst 127:   cmpb    $0, %bl
  Inst 128:   jz      taken=MachLabel(26) not_taken=MachLabel(29)
Block 26:
  (original IR block: block22)
  (successor: Block 27)
  (instruction range: 129 .. 130)
  Inst 129:   jmp     MachLabel(27)
Block 27:
  (successor: Block 28)
  (instruction range: 130 .. 131)
  Inst 130:   jmp     MachLabel(28)
Block 28:
  (original IR block: block20)
  (successor: Block 37)
  (instruction range: 131 .. 132)
  Inst 131:   jmp     MachLabel(37)
Block 29:
  (original IR block: block23)
  (successor: Block 30)
  (instruction range: 132 .. 135)
  Inst 132:   addq    $56, %r14
  Inst 133:   movl    %r13d, %r13d
  Inst 134:   jmp     MachLabel(30)
Block 30:
  (successor: Block 31)
  (instruction range: 135 .. 137)
  Inst 135:   movl    $0, %ebx
  Inst 136:   jmp     MachLabel(31)
Block 31:
  (original IR block: block24)
  (successor: Block 32)
  (successor: Block 33)
  (instruction range: 137 .. 140)
  Inst 137:   movl    0(%r14), %esi
  Inst 138:   cmpl    $0, %esi
  Inst 139:   jz      taken=MachLabel(32) not_taken=MachLabel(33)
Block 32:
  (successor: Block 35)
  (instruction range: 140 .. 141)
  Inst 140:   jmp     MachLabel(35)
Block 33:
  (original IR block: block29)
  (successor: Block 34)
  (instruction range: 141 .. 143)
  Inst 141:   ud2 interrupt
  Inst 142:   jmp     MachLabel(34)
Block 34:
  (successor: Block 35)
  (instruction range: 143 .. 144)
  Inst 143:   jmp     MachLabel(35)
Block 35:
  (original IR block: block30)
  (successor: Block 36)
  (successor: Block 39)
  (instruction range: 144 .. 148)
  Inst 144:   cmpl    %r12d, %ebx
  Inst 145:   setz    %sil
  Inst 146:   cmpb    $0, %sil
  Inst 147:   jnz     taken=MachLabel(36) not_taken=MachLabel(39)
Block 36:
  (successor: Block 37)
  (instruction range: 148 .. 149)
  Inst 148:   jmp     MachLabel(37)
Block 37:
  (original IR block: block19)
  (successor: Block 38)
  (instruction range: 149 .. 150)
  Inst 149:   jmp     MachLabel(38)
Block 38:
  (original IR block: block2)
  (successor: Block 46)
  (instruction range: 150 .. 151)
  Inst 150:   jmp     MachLabel(46)
Block 39:
  (original IR block: block26)
  (successor: Block 40)
  (instruction range: 151 .. 165)
  Inst 151:   movq    %r15/pinned, %rsi // HEAP_BASE
  Inst 152:   addq    %r13, %rsi // r13=bb0714
  Inst 153:   movl    0(%rsi), %esi
  Inst 154:   addl    %ebx, %esi // ebx=2001004
  Inst 155:   movl    %esi, %esi
  Inst 156:   movq    %r15/pinned, %rdi
  Inst 157:   addq    %rsi, %rdi
  Inst 158:   movsbq  0(%rdi), %rdi <-- SEGFAULT here 7ffe543f1004
  Inst 159:   xorl    $-1, %edi
  Inst 160:   movq    %r15/pinned, %rax
  Inst 161:   addq    %rsi, %rax
  Inst 162:   movb    %dil, 0(%rax)
  Inst 163:   addl    $1, %ebx
  Inst 164:   jmp     MachLabel(40)
Block 40:
  (successor: Block 31)
  (instruction range: 165 .. 166)
  Inst 165:   jmp     MachLabel(31)
Block 41:
  (original IR block: block14)
  (successor: Block 42)
  (instruction range: 166 .. 167)
  Inst 166:   jmp     MachLabel(42)
Block 42:
  (successor: Block 45)
  (instruction range: 167 .. 168)
  Inst 167:   jmp     MachLabel(45)
Block 43:
  (original IR block: block13)
  (successor: Block 44)
  (instruction range: 168 .. 169)
  Inst 168:   jmp     MachLabel(44)
Block 44:
  (successor: Block 45)
  (instruction range: 169 .. 171)
  Inst 169:   movl    $1, %edx
  Inst 170:   jmp     MachLabel(45)
Block 45:
  (original IR block: block11)
  (successor: Block 5)
  (instruction range: 171 .. 173)
  Inst 171:   addl    $1, %ecx
  Inst 172:   jmp     MachLabel(5)
Block 46:
  (original IR block: block1)
  (instruction range: 173 .. 176)
  Inst 173:   popq    %r13
  Inst 174:   popq    %r12
  Inst 175:   popq    %rbx
}}
```
</details>

### PR DESCRIPTION
**Pull request by cfallin**
When one branch target label in a MachBuffer is redirected to another,
we eventually fix up branches targetting the first to refer to the
redirected target instead. Separately, we have a branch-folding
optimization that, when an unconditional branch occurs as the only
instruction in a block (right at a label) and the previous instruction
is also an unconditional branch (hence no fallthrough), we can elide
that block entirely and redirect the label. Finally, we prevented
infinite loops when resolving label aliases by chasing only one alias
deep.

Unfortunately, these three facts interacted poorly, and this is a result
of our correctness arguments assuming a fully-general "redirect" that
was not limited to one indirection level. In particular, we could have
some label A that redirected to B, then remove the block at B because it
is just a single branch to C, redirecting B to C. A would still redirect
to B, though, without chasing to C, and hence a branch to B would fall
through to the unrelated block that came after block B.

Thanks to @bnjbvr for finding this bug while debugging the x64 backend
and reducing a failure to the function in issue #2082. (This is a very
subtle bug and it seems to have been quite difficult to chase; my
apologies!)

The fix is to (i) chase redirects arbitrarily deep, but also (ii) ensure
that we do not form a cycle of redirects. The latter is done by very
carefully checking the existing fully-resolved target of the label we
are about to redirect *to*; if it resolves back to the branch that
is causing this redirect, then we avoid making the alias. The comments
in this patch make a slightly more detailed argument why this should be
correct.

Unfortunately we cannot directly test the CLIF that @bnjbvr reduced
because we don't have a way to assert anything about the machine-code
that comes after the branch folding and emission. However, the dedicated
unit tests in this patch replicate an equivalent folding case, and also
test that we handle branch cycles properly (as argued above).

Fixes #2082.

### CODE REVIEW

File: cranelift/codegen/src/machinst/buffer.rs [hunk 1]
```diff
@@ -502,13 +505,11 @@ impl<I: VCodeInst> MachBuffer<I> {
...
 
     /// Resolve a label to an offset, if known. May return `UNKNOWN_LABEL_OFFSET`.
-    fn resolve_label_offset(&self, label: MachLabel) -> CodeOffset {
-        let alias = self.label_aliases[label.0 as usize];
-        if alias != UNKNOWN_LABEL {
-            self.label_offsets[alias.0 as usize]
-        } else {
-            self.label_offsets[label.0 as usize]
+    fn resolve_label_offset(&self, mut label: MachLabel) -> CodeOffset {
+        while self.label_aliases[label.0 as usize] != UNKNOWN_LABEL {
```
Comment (julian-seward1) [inline cranelift/codegen/src/machinst/buffer.rs:509]: Is there anything you can do here to stop this looping infinitely in the case of incorrectly constructed `self.label_aliases`?  On the basis that an assertion failure, in the field, is more informative than a hang.  Like, for example, counting iterations and release-asserting at 1 million?  (I think RA only allows 1 million BBs anyway).

File: cranelift/codegen/src/machinst/buffer.rs [hunk 2]
```diff
@@ -824,6 +825,38 @@ impl<I: VCodeInst> MachBuffer<I> {
...
+                    //
+                    // If L1 were already aliased, through some arbitrarily deep
+                    // chain, to Ln, then it must also resolve to this offset
+                    // `b.start`.
+                    //
+                    // By checking the resolution of `L1` against this offset,
+                    // and aborting this branch-simplification if they are
+                    // equal, we prevent the below assignment from ever creating
+                    // a cycle.
+                    if self.resolve_label_offset(b.target) == b.start {
```
Comment (bnjbvr) [inline cranelift/codegen/src/machinst/buffer.rs:855]: Unless I'm missing something, the exact opposite of this condition is checked right above on line 826, with no apparent side-effects in between, so this is unreachable code, and the existing code already does the right thing. In this case, moving the comment up could be useful.

Comment (cfallin) [reply cranelift/codegen/src/machinst/buffer.rs]: Excellent point, I had missed this. I made a slight change to the existing `if`: we need an `else` that stops the branch-folding loop entirely, to preserve invariants.

File: cranelift/codegen/src/machinst/buffer.rs (hunk 1 above)
Comment (cfallin) [reply cranelift/codegen/src/machinst/buffer.rs:510]: Done!

**Pull request merged**